kube-runtime/src/scheduler.rs

```diff
@@ -152,21 +141,16 @@ where
         let can_take_message = &this.can_take_message;
         let mut scheduler = this.scheduler.as_mut().project();
 
-        while let Poll::Ready(Some(request)) = scheduler.requests.as_mut().poll_next(cx) {
-            scheduler.schedule_message(request);
+        loop {
+            match scheduler.requests.as_mut().poll_next(cx) {
+                Poll::Ready(Some(request)) => scheduler.schedule_message(request),
+                Poll::Ready(None) => return Poll::Ready(None),
+                Poll::Pending => break,
+            }
         }
 
         match scheduler.poll_pop_queue_message(cx, &can_take_message) {
-            Poll::Ready(Some(expired)) => Poll::Ready(Some(expired.context(TimerError))),
-            Poll::Ready(None) => {
-                if scheduler.requests.is_terminated() {
-                    // The source queue has terminated, and all outstanding requests are done, so terminate
-                    Poll::Ready(None)
-                } else {
-                    // The delay queue is empty, but we may get more requests in the future...
-                    Poll::Pending
-                }
```

Comment:
this feels like it's because you've lifted the level of termination handling up towards `on_complete`?

Reply:
This is because of the semantic change from "scheduler terminates when queue + outstanding scheduled items are done" to "scheduler terminates when queue terminates"